ipaserver/install/dogtaginstance.py

```diff
@@ -614,71 +605,158 @@ def _configure_clone(self, subsystem_config, security_domain_hostname,
+        'pki_storage_token',
+        'pki_transport_token',
+    ]
+
+    def __init__(self, subsystem, fqdn, domain, subject_base, ca_subject,
+                 admin_user, admin_password, dm_password,
+                 token_password=None, pki_ini_overlay=None):
+        self.pki_ini_overlay = pki_ini_overlay
+        self.defaults = dict(
             # pretty much static
             ipa_security_domain_name=self.security_domain_name,
-            ipa_ds_database=u"ipaca",
-            ipa_ds_base_dn=self.basedn,
-            ipa_admin_user=self.admin_user,
-            ipa_admin_nickname="ipa-ca-agent",
+            ipa_ds_database=self.ipaca_database,
+            ipa_ds_base_dn=self.ipaca_basedn,
+            ipa_admin_nickname=self.admin_nickname,
             ipa_ca_pem_file=paths.IPA_CA_CRT,
             # variable
-            ipa_ca_subject=self.ca_subject,
-            ipa_subject_base=self.subject_base,
-            ipa_fqdn=self.fqdn,
+            ipa_ca_subject=ca_subject,
+            ipa_subject_base=subject_base,
+            ipa_fqdn=fqdn,
             ipa_ocsp_uri="http://{}.{}/ca/ocsp".format(
-                IPA_CA_RECORD, ipautil.format_netloc(api.env.domain)),
+                IPA_CA_RECORD, ipautil.format_netloc(domain)),
             ipa_admin_cert_p12=paths.DOGTAG_ADMIN_P12,
-            pki_admin_password=self.admin_password,
-            pki_ds_password=self.dm_password,
+            ipa_admin_user=admin_user,
+            pki_admin_password=admin_password,
+            pki_ds_password=dm_password,
             # command line
-            pki_token_password=self.token_password,
+            pki_token_password=token_password,
             # Dogtag's pkiparser defines these config vars by default:
-            pki_dns_domainname=api.env.domain,
-            pki_hostname=self.fqdn,
-            pki_subsystem=self.subsystem.upper(),
-            pki_subsystem_type=self.subsystem.lower(),
+            pki_dns_domainname=domain,
+            pki_hostname=fqdn,
+            pki_subsystem=subsystem.upper(),
+            pki_subsystem_type=subsystem.lower(),
             home_dir=os.path.expanduser("~"),
+            # for softhsm2 testing
+            softhsm2_so=paths.LIBSOFTHSM2_SO
         )
 
-        def mangle_values(d):
-            """Stringify and quote % as %% to avoid interpolation errors
-
-            * booleans are converted to 'True', 'False'
-            * DN and numbers are converted to string
-            * None is turned into empty string ''
-            """
-            result = {}
-            for k, v in d.items():
-                if isinstance(v, (DN, bool, six.integer_types)):
-                    v = six.text_type(v)
-                elif v is None:
-                    v = ''
-                result[k] = v.replace('%', '%%')
-            return result
-
-        defaults = mangle_values(defaults)
-        subsystem_config = mangle_values(subsystem_config)
+    def _mangle_values(self, dct):
```

Comment:
I consider `self` as noise, if it is not used :)  The decorator does improve readability by revealing, at a glance, that the code does not (and cannot) do certain things.

If you dislike the decorator you can promote the function to the module level, which achieves the same readability benefit.

Not a hill I'm going to die on.  I don't mind if we go tit-for-tat with you firmly declining to use `staticmethod` and me subbornly refusing *not* to use it :D

Reply:
For me ``self`` is like the CMB radiation. It's always been there since the big bang of Python. Any time it's missing or deviates from the standard, something special is going on.

In Python ``@classmethod`` is a special case for alternative class constructors or similar methods that don't . ``@staticmethod`` shouldn't be used at all. Static methods were accidentally added, https://mail.python.org/pipermail/python-ideas/2012-May/014969.html

> They're basically an accident -- back in the Python 2.2 days when I was inventing new-style classes and descriptors, I meant to implement class methods but at first I didn't understand them and accidentally implemented static methods first. Then it was too late to remove them and only provide class methods.

I have virtually never used staticmethod and I'm only using classmethod when it's absolutely required.

Reply:
There is a clear readability benefit to annotating that a method does not and cannot access instance variables (`classmethod`) or instance *or* class variables (`staticmethod`).  Readability counts.

But we have been here before.  As I said, it is not a show-stopper for me.